File: packages/javascript/src/llm-derive.ts
import type { GuardEvent } from "./models.js"

/**
 * The developer path (OGR v0.6): a caller forwards the UNTOUCHED provider
 * body — `llm_request` before it reaches the model, `llm_response` before
 * the agent acts on the answer — and the PDP classifies it. This is the
 * SDK-side twin of the hosted runtime's derivation, so the in-process
 * `Runtime` speaks the same contract as the platform.
 *
 * The derivation rewrites the event IN PLACE into the judged shape the
 * detectors read:
 *
 *   llm_request  → `user_input` (a user turn heads the new input) or
 *                  `tool_result` (a continuation feeding outcomes back),
 *                  payload { text?, tool_results?, tools?, system? }
 *   llm_response → `model_output`, payload { text, reasoning?, tool_calls? }
 *
 * ⚠️ Only the NEW input is classified — everything after the last assistant
 * turn. Re-scanning history would double-count findings on every turn.
 *
 * A body matching no known protocol keeps its raw kind with
 * `payload.unparsed = true`: arrived-but-not-judged is a signal, never a
 * silent drop. Protocols: `openai.chat`, `anthropic.messages`;
 * `llm_protocol` is a hint, the shape is sniffed when absent.
 */

type Dict = Record<string, unknown>

const MAX_TOOLS = 64

function asDict(v: unknown): Dict | null {
  return v && typeof v === "object" && !Array.isArray(v) ? (v as Dict) : null
}

function asArray(v: unknown): unknown[] {
  return Array.isArray(v) ? v : []
}

function textOf(content: unknown): string {
  if (typeof content === "string") return content
  return asArray(content)
    .map((part) => {
      const p = asDict(part)
      if (!p) return ""
      if (typeof p.text === "string") return p.text
      if (typeof p.content === "string") return p.content
      return ""
    })
    .filter(Boolean)
    .join("\n")
}

function sniffProtocol(ev: GuardEvent): "openai.chat" | "anthropic.messages" | null {
  const hint = ev.llmProtocol
  if (hint === "openai.chat" || hint === "anthropic.messages") return hint
  const p = ev.payload as Dict
  const messages = asArray(p.messages)
  if (ev.kind === "llm_response") {
    if (asArray(p.choices).length > 0) return "openai.chat"
    if (p.type === "message" || asArray(p.content).length > 0) return "anthropic.messages"
    return null
  }
  if (messages.length === 0) return null
  if (typeof p.system === "string" || typeof p.system === "object") return "anthropic.messages"
  const tools = asArray(p.tools).map(asDict).filter(Boolean) as Dict[]
  if (tools.some((t) => t.input_schema !== undefined)) return "anthropic.messages"
  if (tools.some((t) => t.function !== undefined || t.type === "function")) return "openai.chat"
  const blocks = messages.some((m) =>
    asArray(asDict(m)?.content).some((b) => {
      const t = asDict(b)?.type
      return t === "tool_result" || t === "tool_use"
    }),
  )
  return blocks ? "anthropic.messages" : "openai.chat"
}

function normalizeTools(proto: string, tools: unknown): Dict[] | null {
  const list = asArray(tools).slice(0, MAX_TOOLS).map(asDict).filter(Boolean) as Dict[]
  if (list.length === 0) return null
  return list.map((t) => {
    const fn = asDict(t.function)
    if (proto === "openai.chat" && fn) {
      return { name: fn.name, description: fn.description ?? "", schema: fn.parameters ?? {} }
    }
    return { name: t.name, description: t.description ?? "", schema: t.input_schema ?? t.parameters ?? {} }
  })
}

function deriveRequest(ev: GuardEvent, proto: "openai.chat" | "anthropic.messages"): void {
  const p = ev.payload as Dict
  const messages = asArray(p.messages).map(asDict).filter(Boolean) as Dict[]

  let system = ""
  if (proto === "anthropic.messages") {
    system = typeof p.system === "string" ? p.system : textOf(p.system)
  } else {
    system = messages
      .filter((m) => m.role === "system" || m.role === "developer")
      .map((m) => textOf(m.content))
      .join("\n")
  }

  let lastAssistant = -1
  messages.forEach((m, i) => {
    if (m.role === "assistant") lastAssistant = i
  })
  const fresh = messages.slice(lastAssistant + 1)

  const userTexts: string[] = []
  const results: Dict[] = []
  for (const m of fresh) {
    if (m.role === "tool") {
      results.push({ tool_call_id: m.tool_call_id ?? "", result: textOf(m.content) })
      continue
    }
    if (m.role !== "user") continue
    const blocks = asArray(m.content).map(asDict).filter(Boolean) as Dict[]
    for (const b of blocks.filter((b) => b.type === "tool_result")) {
      results.push({ tool_call_id: b.tool_use_id ?? "", result: textOf(b.content) })
    }
    const text =
      typeof m.content === "string"
        ? m.content
        : blocks
            .filter((b) => b.type === "text")
            .map((b) => String(b.text ?? ""))
            .join("\n")
    if (text) userTexts.push(text)
  }

  const payload: Dict = {}
  const userText = userTexts.join("\n")
  if (userText) payload.text = userText
  if (results.length > 0) payload.tool_results = results
  const tools = normalizeTools(proto, p.tools)
  if (tools) payload.tools = tools
  if (system) payload.system = system

  ev.kind = userText ? "user_input" : results.length > 0 ? "tool_result" : "user_input"
  ev.payload = payload
  ev.llmProtocol = proto
}

function deriveResponse(ev: GuardEvent, proto: "openai.chat" | "anthropic.messages"): void {
  const p = ev.payload as Dict
  const payload: Dict = {}
  const calls: Dict[] = []

  if (proto === "openai.chat") {
    const msg = asDict(asDict(asArray(p.choices)[0])?.message) ?? {}
    payload.text = textOf(msg.content)
    if (typeof msg.reasoning === "string" && msg.reasoning) payload.reasoning = msg.reasoning
    for (const raw of asArray(msg.tool_calls)) {
      const tc = asDict(raw)
      const fn = asDict(tc?.function)
      if (!tc || !fn) continue
      let args: unknown = fn.arguments
      if (typeof args === "string") {
        try {
          args = JSON.parse(args)
        } catch {
          args = { input: args }
        }
      }
      calls.push({ id: tc.id ?? "", name: fn.name ?? "", arguments: args ?? {} })
    }
  } else {
    const blocks = asArray(p.content).map(asDict).filter(Boolean) as Dict[]
    payload.text = blocks
      .filter((b) => b.type === "text")
      .map((b) => String(b.text ?? ""))
      .join("\n")
    const thinking = blocks
      .filter((b) => b.type === "thinking")
      .map((b) => String(b.thinking ?? ""))
      .join("\n")
    if (thinking) payload.reasoning = thinking
    for (const b of blocks) {
      if (b.type !== "tool_use") continue
      calls.push({ id: b.id ?? "", name: b.name ?? "", arguments: b.input ?? {} })
    }
  }

  if (calls.length > 0) payload.tool_calls = calls
  ev.kind = "model_output"
  ev.payload = payload
  ev.llmProtocol = proto
}

/**
 * Classify an `llm_request`/`llm_response` event in place; a no-op for every
 * other kind. Never throws: an unrecognizable body keeps its raw kind with
 * `payload.unparsed = true`.
 */
export function deriveLlmEvent(ev: GuardEvent): void {
  if (ev.kind !== "llm_request" && ev.kind !== "llm_response") return
  try {
    const proto = sniffProtocol(ev)
    if (!proto) {
      ;(ev.payload as Dict).unparsed = true
      return
    }
    if (ev.kind === "llm_request") deriveRequest(ev, proto)
    else deriveResponse(ev, proto)
  } catch {
    ;(ev.payload as Dict).unparsed = true
  }
}

File: packages/python/src/openguardrails/llm_derive.py
"""The developer path (OGR v0.6): classify raw provider bodies.

A caller forwards the UNTOUCHED provider body — ``llm_request`` before it
reaches the model, ``llm_response`` before the agent acts on the answer —
and the PDP classifies it. This is the SDK-side twin of the hosted runtime's
derivation, so the in-process ``Runtime`` speaks the same contract as the
platform.

The derivation rewrites the event IN PLACE into the judged shape the
detectors read:

    llm_request  -> ``user_input`` (a user turn heads the new input) or
                    ``tool_result`` (a continuation feeding outcomes back),
                    payload {text?, tool_results?, tools?, system?}
    llm_response -> ``model_output``, payload {text, reasoning?, tool_calls?}

Only the NEW input is classified — everything after the last assistant turn;
re-scanning history would double-count findings on every turn. A body that
matches no known protocol keeps its raw kind with ``payload["unparsed"] =
True``: arrived-but-not-judged is a signal, never a silent drop. Protocols:
``openai.chat`` and ``anthropic.messages``; ``llm_protocol`` is a hint, the
shape is sniffed when absent.
"""
from __future__ import annotations

import json
from typing import Any

from .models import GuardEvent

_MAX_TOOLS = 64


def _text_of(content: Any) -> str:
    if isinstance(content, str):
        return content
    parts: list[str] = []
    if isinstance(content, list):
        for part in content:
            if not isinstance(part, dict):
                continue
            if isinstance(part.get("text"), str):
                parts.append(part["text"])
            elif isinstance(part.get("content"), str):
                parts.append(part["content"])
    return "\n".join(p for p in parts if p)


def _sniff(ev: GuardEvent) -> str | None:
    if ev.llm_protocol in ("openai.chat", "anthropic.messages"):
        return ev.llm_protocol
    p = ev.payload
    messages = p.get("messages") if isinstance(p.get("messages"), list) else []
    if ev.kind == "llm_response":
        if isinstance(p.get("choices"), list) and p["choices"]:
            return "openai.chat"
        if p.get("type") == "message" or (isinstance(p.get("content"), list) and p["content"]):
            return "anthropic.messages"
        return None
    if not messages:
        return None
    if isinstance(p.get("system"), (str, list)):
        return "anthropic.messages"
    tools = [t for t in (p.get("tools") or []) if isinstance(t, dict)]
    if any("input_schema" in t for t in tools):
        return "anthropic.messages"
    if any("function" in t or t.get("type") == "function" for t in tools):
        return "openai.chat"
    for m in messages:
        for b in m.get("content") if isinstance(m.get("content"), list) else []:
            if isinstance(b, dict) and b.get("type") in ("tool_result", "tool_use"):
                return "anthropic.messages"
    return "openai.chat"


def _normalize_tools(proto: str, tools: Any) -> list[dict] | None:
    items = [t for t in (tools or []) if isinstance(t, dict)][:_MAX_TOOLS]
    if not items:
        return None
    out = []
    for t in items:
        fn = t.get("function") if isinstance(t.get("function"), dict) else None
        if proto == "openai.chat" and fn:
            out.append({"name": fn.get("name"), "description": fn.get("description", ""),
                        "schema": fn.get("parameters", {})})
        else:
            out.append({"name": t.get("name"), "description": t.get("description", ""),
                        "schema": t.get("input_schema", t.get("parameters", {}))})
    return out


def _derive_request(ev: GuardEvent, proto: str) -> None:
    p = ev.payload
    messages = [m for m in (p.get("messages") or []) if isinstance(m, dict)]

    if proto == "anthropic.messages":
        system = p.get("system") if isinstance(p.get("system"), str) else _text_of(p.get("system"))
    else:
        system = "\n".join(_text_of(m.get("content"))
                           for m in messages if m.get("role") in ("system", "developer"))

    last_assistant = -1
    for i, m in enumerate(messages):
        if m.get("role") == "assistant":
            last_assistant = i
    fresh = messages[last_assistant + 1:]

    user_texts: list[str] = []
    results: list[dict] = []
    for m in fresh:
        if m.get("role") == "tool":
            results.append({"tool_call_id": m.get("tool_call_id", ""),
                            "result": _text_of(m.get("content"))})
            continue
        if m.get("role") != "user":
            continue
        blocks = [b for b in (m.get("content") if isinstance(m.get("content"), list) else [])
                  if isinstance(b, dict)]
        for b in blocks:
            if b.get("type") == "tool_result":
                results.append({"tool_call_id": b.get("tool_use_id", ""),
                                "result": _text_of(b.get("content"))})
        text = m["content"] if isinstance(m.get("content"), str) else "\n".join(
            str(b.get("text", "")) for b in blocks if b.get("type") == "text")
        if text:
            user_texts.append(text)

    payload: dict[str, Any] = {}
    user_text = "\n".join(user_texts)
    if user_text:
        payload["text"] = user_text
    if results:
        payload["tool_results"] = results
    tools = _normalize_tools(proto, p.get("tools"))
    if tools:
        payload["tools"] = tools
    if system:
        payload["system"] = system

    ev.kind = "user_input" if user_text or not results else "tool_result"
    ev.payload = payload
    ev.llm_protocol = proto


def _derive_response(ev: GuardEvent, proto: str) -> None:
    p = ev.payload
    payload: dict[str, Any] = {}
    calls: list[dict] = []

    if proto == "openai.chat":
        choices = p.get("choices") or []
        msg = choices[0].get("message", {}) if choices and isinstance(choices[0], dict) else {}
        payload["text"] = _text_of(msg.get("content"))
        if isinstance(msg.get("reasoning"), str) and msg["reasoning"]:
            payload["reasoning"] = msg["reasoning"]
        for tc in msg.get("tool_calls") or []:
            if not isinstance(tc, dict):
                continue
            fn = tc.get("function") if isinstance(tc.get("function"), dict) else {}
            args: Any = fn.get("arguments")
            if isinstance(args, str):
                try:
                    args = json.loads(args)
                except (ValueError, TypeError):
                    args = {"input": args}
            calls.append({"id": tc.get("id", ""), "name": fn.get("name", ""),
                          "arguments": args or {}})
    else:
        blocks = [b for b in (p.get("content") or []) if isinstance(b, dict)]
        payload["text"] = "\n".join(str(b.get("text", "")) for b in blocks
                                    if b.get("type") == "text")
        thinking = "\n".join(str(b.get("thinking", "")) for b in blocks
                             if b.get("type") == "thinking")
        if thinking:
            payload["reasoning"] = thinking
        for b in blocks:
            if b.get("type") == "tool_use":
                calls.append({"id": b.get("id", ""), "name": b.get("name", ""),
                              "arguments": b.get("input", {})})

    if calls:
        payload["tool_calls"] = calls
    ev.kind = "model_output"
    ev.payload = payload
    ev.llm_protocol = proto


def derive_llm_event(ev: GuardEvent) -> None:
    """Classify an ``llm_request``/``llm_response`` event in place; a no-op
    for every other kind. Never raises: an unrecognizable body keeps its raw
    kind with ``payload["unparsed"] = True``."""
    if ev.kind not in ("llm_request", "llm_response"):
        return
    try:
        proto = _sniff(ev)
        if proto is None:
            ev.payload["unparsed"] = True
            return
        if ev.kind == "llm_request":
            _derive_request(ev, proto)
        else:
            _derive_response(ev, proto)
    except Exception:  # noqa: BLE001 — honesty over crashing the decision path
        ev.payload["unparsed"] = True
